math/src/polynom/univariate.rs

```diff
@@ -0,0 +1,263 @@
+use alloc::vec::Vec;
+
+use utils::{ByteReader, ByteWriter, Deserializable, DeserializationError, Serializable};
+
+use crate::{batch_inversion, polynom, FieldElement};
+
+// COMPRESSED UNIVARIATE POLYNOMIAL
+// ================================================================================================
+
+/// The coefficients of a univariate polynomial of degree n with the linear term coefficient
+/// omitted.
+///
+/// This compressed representation is useful during the sum-check protocol as the full uncompressed
+/// representation can be recovered from the compressed one and the current sum-check round claim.
+#[derive(Clone, Debug)]
+pub struct CompressedUnivariatePolyCoef<E: FieldElement>(Vec<E>);
+
+impl<E: FieldElement> CompressedUnivariatePolyCoef<E> {
+    /// Evaluates a polynomial at a challenge point using a round claim.
+    ///
+    /// The round claim is used to recover the coefficient of the linear term using the relation
+    /// 2 * c0 + c1 + ... c_{n - 1} = claim. Using the complete list of coefficients, the polynomial
+    /// is then evaluated using Horner's method.
+    pub fn evaluate_using_claim(&self, claim: &E, challenge: &E) -> E {
+        // recover the coefficient of the linear term
+        let c1 = *claim - self.0.iter().fold(E::ZERO, |acc, term| acc + *term) - self.0[0];
+
+        // construct the full coefficient list
+        let mut complete_coefficients = vec![self.0[0], c1];
+        complete_coefficients.extend_from_slice(&self.0[1..]);
+
+        // evaluate
+        polynom::eval(&complete_coefficients, *challenge)
+    }
+}
+
+impl<E: FieldElement> Serializable for CompressedUnivariatePolyCoef<E> {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        self.0.write_into(target);
+    }
+}
+
+impl<E> Deserializable for CompressedUnivariatePolyCoef<E>
+where
+    E: FieldElement,
+{
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        Ok(Self(Deserializable::read_from(source)?))
+    }
+}
+
+/// The evaluations of a univariate polynomial of degree n at 0, 1, ..., n with the evaluation at 0
+/// omitted.
+///
+/// This compressed representation is useful during the sum-check protocol as the full uncompressed
+/// representation can be recovered from the compressed one and the current sum-check round claim.
+#[derive(Clone, Debug)]
+pub struct CompressedUnivariatePolyEvals<E>(Vec<E>);
```

Comment:
Similar to one of the previous comments: I wonder if we should use `SmallVec` here instead of `Vec`.

Reply:
Changed